sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { PipelineResponse, PipelineRequest, SendRequest } from "../interfaces";
+import { PipelinePolicy } from "../pipeline";
+import { delay } from "../util/helpers";
+import { createClientLogger } from "@azure/logger";
+import { RetryStrategy, RetryStrategyState } from "../retryStrategies/retryStrategy";
+import { RestError } from "../restError";
+import { AbortError } from "@azure/abort-controller";
+
+const retryPolicyLogger = createClientLogger("core-rest-pipeline retryPolicy");
+const DEFAULT_MAX_RETRIES = 3;
+
+/**
+ * The programmatic identifier of the retryPolicy.
+ */
+const retryPolicyName = "retryPolicy";
+
+/**
+ * retryPolicy is a generic policy to enable retrying requests when certain conditions are met
+ */
+export function retryPolicy(...strategies: RetryStrategy[]): PipelinePolicy {
+  return {
+    name: retryPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      let response: PipelineResponse | undefined;
+      let responseError: RestError | undefined;
+      let retryCount = -1;
+
+      const strategyState: RetryStrategyState[] = Array(strategies.length);
+
+      // eslint-disable-next-line no-constant-condition
+      retryRequest: while (true) {
+        retryCount += 1;
+        response = undefined;
+        responseError = undefined;
+
+        try {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Attempting to send request`,
+            request.requestId
+          );
+          response = await next(request);
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received a response from request`,
+            request.requestId
+          );
+        } catch (e) {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received an error from request`,
+            request.requestId
+          );
+          responseError = e as RestError;
+          if (responseError.name !== "RestError") {
+            throw responseError;
+          }
+          response = responseError.response;
+        }
+
+        retryPolicyLogger.info(
+          `Retry ${retryCount}: Processing ${strategies.length} retry strategies.`
+        );
+
+        for (const [i, strategy] of strategies.entries()) {
+          const strategyLogger = strategy.logger || retryPolicyLogger;
+          strategyLogger.info(`Retry ${retryCount}: Processing retry strategy ${strategy.name}.`);
+          let state: RetryStrategyState = {
+            ...strategyState[i],
+            retryCount,
+            response,
+            responseError
+          };
+
+          if (state.retryCount >= (state.maxRetries || DEFAULT_MAX_RETRIES)) {
```

Comment:
I am curious why we are using the `retryCount` from the strategy state, rather than the local variable `retryCount` that this policy is keeping. Presumably the strategy could do whatever it wanted to manipulate this value, but I'm not sure why it would differ?

Similarly, it feels a bit odd that the strategy gets to reason about max retries, presumably that was so the outer policy wouldn't have to. But at this point, why have the policy reason about retryCount at all? Whatever strategy was persisting retryCount and consulting maxRetries could just raise an error / return the response?

I have a small concern that this is putting too much pressure on exponentialRetryStrategy. What if I want to use this unified retry policy *without* using exponential retries, but I still want the other two strategies, and I want to use a max retry count that *isn't* the default? It doesn't seem possible currently.

Reply:
`retryCount` is useful information for strategies, I think we can agree on that.
- Strategies might want to know how many reties have been running so far.
- This is similar to mapping and iterating functions. Not providing an index means developers are forced to fabricate their own indexes if needed, which seems unfortunate.

Whether `retryCount` should be part of the `state` is more about how we can split the state into inputs vs outputs:
-  We could have two types, one corresponding to the parameters to the strategy methods, and the other type corresponding to the possible responses. Would that help?

> I have a small concern that this is putting too much pressure on exponentialRetryStrategy.

This line isn’t tied at all to `exponentialRetryStrategy`, other than the fact that exponentialRetryStrategy uses the state.

> What if I want to use this unified retry policy without using exponential retries, but I still want the other two strategies, and I want to use a max retry count that isn't the default? It doesn't seem possible currently.

Something like this:

```
retryPolicy(
  {
    name,
    updateRetryState: state => {
      …state,
      maxRetries: customMaxRetries
    }
  },
  strategy1(),
  strategy2()
)

Reply:
It feels a bit strange to need to invent a strategy just to pass max retries along, I'm not sure how someone would easily determine that from the public apis provided by these policies

Reply:
How does this look? https://github.com/Azure/azure-sdk-for-js/pull/19078/commits/706de41ce49045e4afaa8559dafcd8204a1e122c